src/main/java/ru/yandex/clickhouse/response/parser/ClickHouseDateValueParser.java

```diff
@@ -0,0 +1,175 @@
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.time.format.ResolverStyle;
+import java.util.Objects;
+import java.util.TimeZone;
+import java.util.regex.Pattern;
+
+import ru.yandex.clickhouse.except.ClickHouseException;
+import ru.yandex.clickhouse.except.ClickHouseUnknownException;
+import ru.yandex.clickhouse.response.ByteFragment;
+import ru.yandex.clickhouse.response.ClickHouseColumnInfo;
+
+abstract class ClickHouseDateValueParser<T> extends ClickHouseValueParser<T> {
+
+    private static final Pattern PATTERN_EMPTY_DATE =
+        Pattern.compile("^(0000-00-00|0000-00-00 00:00:00|0)$");
+
+    private static final DateTimeFormatter DATE_FORMATTER =
+        DateTimeFormatter.ofPattern("yyyy[-]MM[-]dd");
+    private static final DateTimeFormatter DATE_TIME_FORMATTER =
+        DateTimeFormatter.ofPattern("yyyy-MM-dd['T'][ ]HH:mm:ss[.SSS]");
+    private static final DateTimeFormatter TIME_FORMATTER_NUMBERS =
+        DateTimeFormatter.ofPattern("HH[mm][ss]")
+            .withResolverStyle(ResolverStyle.STRICT);
+
+    private final Class<T> clazz;
+
+    protected ClickHouseDateValueParser(Class<T> clazz) {
+        this.clazz = Objects.requireNonNull(clazz);
+    }
+
+    @Override
+    public T parse(ByteFragment value, ClickHouseColumnInfo columnInfo,
+        TimeZone timeZone) throws ClickHouseException
+    {
+
+        if (value.isNull()) {
+            return null;
+        }
+
+        String s = value.asString();
+
+        /*
+         *  filter default values for relevant data types,
+         *  even if the column has nullable flag set.
+         */
+        if (PATTERN_EMPTY_DATE.matcher(s).matches()) {
+            return null;
+        }
+
+        switch (columnInfo.getEffectiveClickHouseDataType()) {
+            case Date:
+                try {
+                    return parseDate(s, columnInfo, timeZone);
+                } catch (Exception e) {
+                    throw new ClickHouseUnknownException(
+                        "Error parsing '" + s + "' of data type '"
+                            + columnInfo.getOriginalTypeName()
+                            + "' as " + clazz.getName(),
+                        e);
+                }
+            case DateTime:
+                try {
+                    return parseDateTime(s, columnInfo, timeZone);
+                } catch (Exception e) {
+                    throw new ClickHouseUnknownException(
+                        "Error parsing '" + s + "' of data type '"
+                            + columnInfo.getOriginalTypeName()
+                            + "' as " + clazz.getName(),
+                        e);
+                }
+            case Int8:
+            case Int16:
+            case Int32:
+            case Int64:
+            case UInt8:
+            case UInt16:
+            case UInt32:
+            case UInt64:
+                try {
+                    long l = Long.parseLong(s);
```

Comment:
1. We may have UInt64 overflow here
2. Why all these things in class `ClickHouseDateValueParser`? That's really confusing.

Reply:
1. Thanks, I had not thought of this. New implementation truncates to milliseconds.
2. The code should not be confusing, let's see if we can improve it.  `ClickHouseDateValueParser` provides basic functionality for all date/time parsing. This makes the individual implementations much easier to read. We _could_ move the "utility" methods (lines 146 ff.) into the actual implementations. Do you think that this would help overall simplicity?